Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -4,34 +4,231 @@
 
 #define SUCCESS 0
 #define ERROR -1
-#define OVERFLOW -2
-#define EMPTY -3
+
+#define TEST 9
+#define EXPECTED 10
+
+#define SIZE 10
 
 // Hash struct
-typedef struct Hash {
+typedef struct Item {
     int key;
     int value;
-} Hash;
+} Item;
 
-void print() {
+//create global testHash and expectedHash for testFunctions to use
+Item* testHash[SIZE];
+Item* expectedHash[SIZE];
 
+// traverse function is for visual confirmation of differences between hashTables when building functionality. 
+// void traverse() {
+//     for (int i = 0; i < SIZE; i++) {
+//         int value = testHash[i]->value;
+//         int key = testHash[i]->key;        
+//         printf("Test Key: %d Value: %d \n", key, value);
+//     }
+//     for (int i = 0; i < SIZE; i++) {
+//         int value = expectedHash[i]->value;
+//         int key = expectedHash[i]->key;
+//         printf("Expected Key: %d Value: %d \n", key, value);
+//     }
+// }
+
+//compare checks the testHash and expectedHash for parity
+// if the two HashTables match returns true
+// if a mismatch occurs returns false
+bool compare() {
+    for (int i = 0; i < SIZE; i++) {
+        int testKey = testHash[i]->key;
+        int testValue = testHash[i]->value;
+        int expectedKey = expectedHash[i]->key;
+        int expectedValue = expectedHash[i]->value;
+        if (testKey != expectedKey || testValue != expectedValue) {
+            return false;
+        }
+    }
+    return true;
 }
 
-void traverse() {
+// hashing takes in a key
+// returns a "hashed" index based on the SIZE of the hashTable
+int hashing(int key) {
+    return key % SIZE;
+}
 
+// creates a new Item struct
+// returns new struct
+Item * createItem(int key, int value, int success) {
+    Item * item = malloc(sizeof(Item));
+    if (item == NULL) {
+    } else {
+        item->key = key;
+        item->value = value;
+        success = SUCCESS;
+        return item;
+    }
 }
 
-bool compare() {
-    return true;
-    return false;
+// insert takes in a key and value
+// calls a helper function to create an Item struct with the provided key and value
+// adds the new Item to the hashTable
+// if the expected value is in the hashTable at the expected key then returns a success code
+void insert(int type, int key, int value, int success) {
```

Comment:
I'd expect insert to take a pointer to the hash table as the first parameter.  See the api for https://developer.gnome.org/glib/2.28/glib-Hash-Tables.html#g-hash-table-insert.  This will also help you eliminate the global arrays.

I'm not sure what @klizhentas has asked you to do, but I'd also consider implementing at string keys and values, such that you could unpack a json dict.  In terms of teaching pointer manipulation, this will go a lot further than hashing ints.  E.g. what would need to change to go from being able to hash:

> {1: 2}

To being able to hash:

> {"foo": "bar"}

Reply:
I agree with Walt, let's implement "string" -> int hash table, mostly because strings are much more interesting scenario

Reply:
Can you provide some clarity - is the expectation that the hashTable should be flexible enough to accept an int AND a char array/string as a key or the char arry/string instead of the current int? Thank you!

Reply:
I'd say that 'key: string' -> 'value: int' is sufficient.  While handling arbitrary input keys is interesting, you'll have gotten 80% of the value by generalizing only to strings.  Plus, with the current revision you already showed int inputs.

Reply:
Addressed in [commit 1aae87af1f3359951eb45fafea44c753b86ed852](https://github.com/kcarrel/Training-Notes/pull/12/commits/1aae87af1f3359951eb45fafea44c753b86ed852)!